vald/evm/evm.go

```diff
@@ -486,6 +467,48 @@ func (mgr Mgr) GetTxReceiptIfFinalized(chain nexus.ChainName, txID common.Hash,
 	return txReceipt, nil
 }
 
+// GetTxReceiptsIfFinalized retrieves receipts for provided transaction IDs, only if they're finalized.
+func (mgr Mgr) GetTxReceiptsIfFinalized(chain nexus.ChainName, txIDs []common.Hash, confHeight uint64) ([]rs.Result[*geth.Receipt], error) {
+	client, ok := mgr.rpcs[strings.ToLower(chain.String())]
+	if !ok {
+		return nil, fmt.Errorf("rpc client not found for chain %s", chain.String())
+	}
+
+	results, err := client.TransactionReceipts(context.Background(), txIDs)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(errors.With(err, "chain", chain.String()), "cannot get transaction receipts")
+	}
+
+	isFinalized := func(r rpc.Result) rs.Result[*geth.Receipt] {
+		result := rs.Result[*geth.Receipt](r)
+
+		// Treat non-existent tx or tx in the mempool as not finalized.
+		if goerrors.Is(result.Err(), ethereum.NotFound) {
+			return rs.FromErr[*geth.Receipt](NotFinalized)
```

Comment:
for debugging purposes, it's useful to distinguish between tx not found vs tx found and not finalized. I'd prefer separate errors or another approach here

Reply:
technically non-existent txs or txs in the mempool are not finalized, but if you have strong opinion on this, I can distinguish.